clients/drcachesim/tracer/output.cpp

```diff
@@ -979,19 +979,22 @@ process_and_output_buffer(void *drcontext, bool skip_size_cap)
     if (data->has_thread_header && op_offline.get_value())
         header_size += data->init_header_size;
 
-    if (align_attach_detach_endpoints()) {
-        // This is the attach counterpart to instru_t::set_frozen_timestamp(): we place
-        // timestamps at buffer creation, but that can be before we're fully attached.
-        // We update any too-early timestamps to reflect when we actually started
-        // tracing.  (Switching back to timestamps at buffer output is actually
-        // worse as we then have the identical frozen timestamp for all the flushes
-        // during detach, plus they are all on the same cpu too.)
-        uint64 min_timestamp = attached_timestamp.load(std::memory_order_acquire);
-        if (min_timestamp == 0) {
-            // This data is too early: we drop it.
-            NOTIFY(1, "Dropping too-early data for T%zd\n", dr_get_thread_id(drcontext));
-            BUF_PTR(data->seg_base) = data->buf_base + header_size;
-            return;
+    if (align_attach_detach_endpoints() || has_tracing_windows()) {
```

Comment:
So now if windows are enabled and we attach to the process, we'll align the buffers at the start even if -align_endpoints is explicitly turned off by the user.  I don't think that's what we want.  Maybe it is best to have a separate timestamp for restarting tracing (see same comment elsewhere).

Reply:
If I understand where the aligning logic is, we only only align (where it drops the too-early data) when `align_attach_detach_endpoints()` is true from the inner `if` on line 984. When `has_tracing_windows()` is true, we just refresh.

Reply:
No, the `if` on line 984 doesn't do anything unless min_timestamp is 0.  The refresh on line 1001 will be called even if -align_endpoints is off.

Reply:
I see. Are we ok with re-refreshing when we have both `-align_endpoings` and `has_tracing_windows()`? This is what I am thinking. Keep the original local for refreshing the header timestamp and add a new refresh with `retrace_attached_timestamp` when `has_tracing_windows()`:

```
    size_t stamp_offs =
        header_size > buf_hdr_slots_size ? header_size - buf_hdr_slots_size : 0;
    uint64 min_timestamp;
    if (align_attach_detach_endpoints()) {
        // This is the attach counterpart to instru_t::set_frozen_timestamp(): we
        // place timestamps at buffer creation, but that can be before we're fully
        // attached. We update any too-early timestamps to reflect when we actually
        // started tracing.  (Switching back to timestamps at buffer output is
        // actually worse as we then have the identical frozen timestamp for all the
        // flushes during detach, plus they are all on the same cpu too.)
        min_timestamp = attached_timestamp.load(std::memory_order_acquire);
        if (min_timestamp == 0) {
            // This data is too early: we drop it.
            NOTIFY(0, "Dropping too-early data for T%zd\n",
                   dr_get_thread_id(drcontext));
            BUF_PTR(data->seg_base) = data->buf_base + header_size;
            return;
        }
        instru->refresh_unit_header_timestamp(data->buf_base + stamp_offs, min_timestamp);
    }

    if (has_tracing_windows()) {
        min_timestamp = retrace_attached_timestamp.load(std::memory_order_acquire);
        instru->refresh_unit_header_timestamp(data->buf_base + stamp_offs, min_timestamp);
    }
```

Reply:
Well, it only changes the timestamp if it's less than the passed-in min.  Maybe it should be renamed to s/refresh/clamp/ to make its conditional nature more clear.

Reply:
Done. Added `retrace_attached_timestamp` and `s/refresh/clamp/ `. Now `retrace_attached_timestamp` is updated when we start retracing and is used "clamp" the header timestamp when `has_tracing_windows()`

Reply:
Since we don't do a detach and (re-)attach and DR was attached the entire time it may be best to not have "attached" in the name.  `retrace_start_timestamp` or something?